include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h

```diff
@@ -122,31 +210,17 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
     constexpr ::std::size_t elems_per_workitem = _KernelParam::elems_per_workitem;
 
     // Avoid non_uniform n by padding up to a multiple of wgsize
-    std::uint32_t elems_in_tile = wgsize * elems_per_workitem;
+    ::std::uint32_t elems_in_tile = wgsize * elems_per_workitem;
     ::std::size_t num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(n, elems_in_tile);
     ::std::size_t num_workitems = num_wgs * wgsize;
 
-    constexpr int status_flag_padding = SUBGROUP_SIZE;
-    std::size_t status_flags_elems = num_wgs + status_flag_padding + 1;
-    std::size_t status_flags_size = status_flags_elems * sizeof(std::uint32_t);
-
-    std::size_t tile_sums_elems = num_wgs + status_flag_padding;
-    std::size_t tile_sums_size = status_flags_elems * sizeof(_Type);
-
-    std::size_t extra_mem_for_aligment = alignof(_Type) - (status_flags_size % alignof(_Type));
-    // status_flags_size for the status_flags
-    // extra_mem_for_aligment of the datatype _Type
-    // First tile_sums_size partial scanned values
-    // Second tile_sums_size full scanned values (current partial plus all previous workgroups partial)
-    char* mem_pool =
-        sycl::malloc_device<char>(status_flags_size + extra_mem_for_aligment + 2 * tile_sums_size, __queue);
-
-    std::size_t tile_sums_offset = status_flags_size + extra_mem_for_aligment;
-
-    std::uint32_t* status_flags = reinterpret_cast<std::uint32_t*>(mem_pool);
-    _Type* tile_sums = reinterpret_cast<_Type*>(mem_pool + tile_sums_offset);
+    char* scratch = scan_memory<_Type>::allocate_memory(num_wgs, __queue);
 
+    // TODO: can we get rid of this section, i.e. hide the dynamic_tile_id_counter, hide initialization details
```

Comment:
Sent you this offline, but for the record, we could copy some of [this](https://github.com/oneapi-src/oneDPL/blob/f132aa20f0d0b7f02f59cbd2d93a9f680369f228/include/oneapi/dpl/experimental/kt/internal/gpu_radix_sort_dispatchers.h#L134) functionality.

Reply:
Yeah. I've changed a bit how we do this, but I am still not totally convinced. I think the sensible thing would be to reestructure memory within the `scratch`, and instead of having `[Flags, Counter, Values]` we should move things around to have `[Values, Flags, Counter]`.
This leaves the initialization as is, only needing to change were is the starting address, while removing the counter from the `Scan_memory` which should be contiguous. 

However, this PR is big enough as is, and that change should be submitted in  a different PR IMO

Reply:
As we've agreed in our other conversation, this  changes are also included in this PR.